### DISCUSSION TITLE
can't install python by pyenv

### DISCUSSION THREAD

**Discussion by kev26** [pyenv/pyenv#2401] [Q&A] [answered]
I'm new learner linux, so when I trying to install python via pyenv, it appear this error :
OS: Ubuntu 22.04
```
kev@kev-Modern-15-A5M:~$ pyenv install 3.8.10
Downloading Python-3.8.10.tar.xz...
-> https://www.python.org/ftp/python/3.8.10/Python-3.8.10.tar.xz
Installing Python-3.8.10...
patching file Misc/NEWS.d/next/Build/2021-10-11-16-27-38.bpo-45405.iSfdW5.rst
patching file configure
patching file configure.ac
WARNING: The Python bz2 extension was not compiled. Missing the bzip2 lib?
WARNING: The Python readline extension was not compiled. Missing the GNU readline lib?
ERROR: The Python ssl extension was not compiled. Missing the OpenSSL lib?

Please consult to the Wiki page to fix the problem.
https://github.com/pyenv/pyenv/wiki/Common-build-problems


BUILD FAILED (Ubuntu 22.04 using python-build 20180424)

Inspect or clean up the working tree at /tmp/python-build.20220629154809.158632
Results logged to /tmp/python-build.20220629154809.158632.log

Last 10 log lines:
	 ./python -E -m ensurepip \
		$ensurepip --root=/ ; \
fi
Looking in links: /tmp/tmpdi0xus81
Processing /tmp/tmpdi0xus81/setuptools-56.0.0-py3-none-any.whl
Processing /tmp/tmpdi0xus81/pip-21.1.1-py3-none-any.whl
Installing collected packages: setuptools, pip
  WARNING: The scripts pip3 and pip3.8 are installed in '/home/kev/.pyenv/versions/3.8.10/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed pip-21.1.1 setuptools-56.0.0
kev@kev-Modern-15-A5M:~$ 

```

**Comment by hitblast** [comment]
It appears that you are missing **OpenSSL** in your development environment, and since it is a core dependency for running pyenv, you will need to install it first.

On Ubuntu, you can run the following commands to install the needed packages. More of these instructions can be found in the [Install Python build dependencies](https://github.com/pyenv/pyenv#install-python-build-dependencies) section of the included README.md file.

Copy-paste these commands into your terminal and run:

```bash
sudo apt-get update; sudo apt-get install make build-essential libssl-dev zlib1g-dev \
libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm \
libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev
```

**Reply by kev26** [reply]
I had installed OpenSSL, but seem because I installed it via Hombrew, and is this reason cause the error ?
```
kev@kev-Modern-15-A5M:~$ openssl version -a
OpenSSL 1.1.1p  21 Jun 2022
built on: Tue Jun 21 13:39:39 2022 UTC
platform: linux-x86_64
options:  bn(64,64) md2(char) rc4(8x,int) des(int) idea(int) blowfish(ptr) 
compiler: gcc-5 -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAESNI_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPOLY1305_ASM -DNDEBUG
OPENSSLDIR: "/home/linuxbrew/.linuxbrew/etc/openssl@1.1"
ENGINESDIR: "/home/linuxbrew/.linuxbrew/Cellar/openssl@1.1/1.1.1p/lib/engines-1.1"
```
So how to set path to OpenSSL in hombrew when I install python via pyenv ? Thank you !!!

**Comment by kev26** [comment]
I tried to uninstall OpenSSL in Hombrew and try to installed it by `sudo apt install openssl`, it installed ok, but when I check openssl version and its path, seem the path still forced with Hombew :
```
kev@kev-Modern-15-A5M:/$ openssl version -a
bash: /home/linuxbrew/.linuxbrew/bin/openssl: No such file or directory
```
So how to set path to Openssl in Homebrew when I install python via pyenv or how to set default path in system instead of Homebew ?

**Reply by kev26** [reply]
thank you so much, and I don't need to use Homebrew to install dependencies anymore? it's more of an inconvenience than a good utility!

**Reply by hitblast** [reply]
@kev26 Well, it will totally depend on you whether you want to install dependencies with it or not. It's still an option :)

**Comment by hitblast** [comment, marked as answer]
Just quote-replying again if someone else needs answer to the same issue.

> I tried to uninstall OpenSSL in Hombrew and try to installed it by `sudo apt install openssl`, it installed ok, but when I check openssl version and its path, seem the path still forced with Hombew :
> 
> ```
> kev@kev-Modern-15-A5M:/$ openssl version -a
> bash: /home/linuxbrew/.linuxbrew/bin/openssl: No such file or directory
> ```
> 
> So how to set path to Openssl in Homebrew when I install python via pyenv or how to set default path in system instead of Homebew ?

Okay, so assuming you've uninstalled OpenSSL from **Homebrew** and reinstalled it using **Apt**, you can try running the following commands in order to remove any possible symlinks that originated from the previous installation of OpenSSL and check if it redirects to the current installation. 

```bash
brew cleanup && which openssl
```

---

Also to clean up the PATH variable and remove the previous entries of OpenSSL, you can first view the `.bashrc` file to check if there is anything relatable. For instance, if there is something like this inside

```bash
export PATH=$PATH:/home/linuxbrew/.linuxbrew/bin/openssl
```

then you must remove it save the file and restart your terminal session. Same goes for any non-interactive shell file (`.bash_profile` or anything else). I hope this helps.
